### PR TITLE
[WIP] OpenACC back-end

### PR DESCRIPTION
**Pull request by BenjaminW3**
Some first tests for OpenACC support for #248.

### CODE REVIEW

**Comment by theZiz** [comment]
So, what is missing right now before we can merge OpenACC?

**Comment by BenjaminW3** [comment]
First of all, there are compile errors e.g. `ExecCpuOpenAcc2.hpp:194:79: error: ‘gridBlockExtent’ referenced in target region does not have a mappable type`.
Secondly, I was unable to find a way to explicitly synchronize threads within a block. If this is not possible, we would have to add the restriction, that the blockThreadExtend can only ever be 1 (similar to the TBB executor).
Last but not least, the current code directly accesses host memory so we can only use this as non-offloading accelerator (similar to the OpenMP4 back-end).

**Comment by psychocoderHPC** [comment]
> Secondly, I was unable to find a way to explicitly synchronize threads within a block. If this is not possible, we would have to add the restriction, that the blockThreadExtend can only ever be 1 (similar to the TBB executor).

To set blockThreadExtend force to one is fine on a CPU but on a GPU

**Comment by psychocoderHPC** [comment]
A short internal discussion with @theZiz pointed out that we should limit for the first kick of of this back-end the number of threads per block to one.

> Last but not least, the current code directly accesses host memory so we can only use this as non-offloading accelerator (similar to the OpenMP4 back-end)

We need to check if we can workaround this, else we will not fulfill our goal to have a single source code.

> First of all, there are compile errors e.g. ExecCpuOpenAcc2.hpp:194:79: error: ‘gridBlockExtent’ referenced in target region does not have a mappable type.

This needs further investigation

**Comment by BenjaminW3** [comment]
CMake 3.10 adds his own `FindOpenACC.cmake`. So the one currently included can be removed.

File: cmake/modules/FindOpenACC.cmake [hunk 1]
```diff
@@ -0,0 +1,104 @@
+#.rst:
```
Comment (BenjaminW3) [inline cmake/modules/FindOpenACC.cmake]: CMake 3.10 has added a FindOpenACC script so this should not be necessary anymore.

**Comment by ax3l** [comment]
From the PIConGPU side it's ok if Alpaka moves to CMake 3.10+ as well.

**Comment by ax3l** [comment]
We are members in the OpenACC org and regularly at user group meetings (e.g. the next at GTC). Can we formulate the two main points,

> Secondly, I was unable to find a way to explicitly synchronize threads within a block.

and

> the current code directly accesses host memory so we can only use this as non-offloading accelerator (similar to the OpenMP4 back-end).

as a proposal to the [OpenACC specification](https://www.openacc.org/resources)? What would you like to have added? I think we did let these two points to long at rest without actively proposing changes we need to the OpenACC community and should do this.

File: cmake/modules/FindOpenACC.cmake (hunk 1 above)
Comment (ax3l) [reply cmake/modules/FindOpenACC.cmake]: From the PIConGPU side it's ok if Alpaka moves to CMake 3.10+ as well.

**Comment by ax3l** [comment]
> the current code directly accesses host memory so we can only use this as non-offloading accelerator (similar to the OpenMP4 back-end).

@BenjaminW3 is this a limitation of OpenACC or just not implemented in the PR?

**Comment by BenjaminW3** [comment]
@ax3l I do not remember anymore. It is too long ago.
If OpenACC has a method to allocate memory on a device from host code, it should be possible. I do not know if this exists.

File: include/alpaka/acc/AccCpuOpenAcc2.hpp [hunk 2]
```diff
@@ -0,0 +1,307 @@
+/**
+* \file
+* Copyright 2018 Benjamin Worpitz
+*
+* This file is part of alpaka.
+*
+* alpaka is free software: you can redistribute it and/or modify
```
Comment (ax3l) [inline include/alpaka/acc/AccCpuOpenAcc2.hpp:7]: With regards to #637, this new file's header will need an update :)

File: include/alpaka/atomic/AtomicOpenAcc.hpp [hunk 3]
```diff
@@ -0,0 +1,371 @@
+/**
+* \file
+* Copyright 2018 Benjamin Worpitz
+*
+* This file is part of alpaka.
+*
+* alpaka is free software: you can redistribute it and/or modify
```
Comment (ax3l) [inline include/alpaka/atomic/AtomicOpenAcc.hpp:7]: With regards to #637, this new file's header will need an update :)

File: include/alpaka/block/shared/dyn/BlockSharedMemDynRef.hpp [hunk 4]
```diff
@@ -0,0 +1,89 @@
+/**
+* \file
+* Copyright 2018 Benjamin Worpitz
+*
+* This file is part of alpaka.
+*
+* alpaka is free software: you can redistribute it and/or modify
```
Comment (ax3l) [inline include/alpaka/block/shared/dyn/BlockSharedMemDynRef.hpp:7]: With regards to #637, this new file's header will need an update :)

File: include/alpaka/exec/ExecCpuOpenAcc2.hpp [hunk 5]
```diff
@@ -0,0 +1,337 @@
+/**
+* \file
+* Copyright 2018 Benjamin Worpitz
+*
+* This file is part of alpaka.
+*
+* alpaka is free software: you can redistribute it and/or modify
```
Comment (ax3l) [inline include/alpaka/exec/ExecCpuOpenAcc2.hpp:7]: With regards to #637, this new file's header will need an update :)

File: include/alpaka/idx/bt/IdxBtRef.hpp [hunk 6]
```diff
@@ -0,0 +1,129 @@
+/**
+* \file
+* Copyright 2018 Benjamin Worpitz
+*
+* This file is part of alpaka.
+*
+* alpaka is free software: you can redistribute it and/or modify
```
Comment (ax3l) [inline include/alpaka/idx/bt/IdxBtRef.hpp:7]: With regards to #637, this new file's header will need an update :)

File: cmake/modules/FindOpenACC.cmake (hunk 1 above)
Comment (ax3l) [reply cmake/modules/FindOpenACC.cmake:1]: With regards to #637, this new file will need an MPL-2.0 header :)

**Pull request closed without merging**